### Data Cleaning in Python
###    1. Dropping Columns in a DataFrame    
###    2. Changing the index of a DataFrame.
###    3. Tidying up Fields in the Data
###    4. Combining str Methods with NumPy to Clean Columns
###    5. Renaming Columns and Skipping Rows


### Import the required modules

In [1]:
import pandas as pd
import numpy as np

### 1. Dropping Columns in a DataFrame

### Often, you’ll find that not all the categories of data in a dataset are useful to you. 
### For example, you might have a dataset containing student information (name, grade, standard, parents’ names, and address) but want to focus on analyzing student grades.

### In this case, the address or parents’ names categories are not important to you. Retaining these unneeded categories will take up unnecessary space and potentially also bring down runtime.

### Pandas provides a handy way of removing unwanted columns or rows from a DataFrame with the drop() function. 
### First, let’s create a DataFrame out of the CSV file ‘BL-Flickr-Images-Book.csv’. 
### In the examples below, we pass an absolute path to pd.read_csv.

In [8]:
df = pd.read_csv('/home/lenovo/Documents/Course Material 2020-21/Data Analytics 2020-21/Datasets for DA/BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

### Some columns provide ancillary information that would be helpful to the library but isn’t very descriptive of the books themselves: Edition Statement, Corporate Author, Corporate Contributors, Former owner, Engraver, Issuance type and Shelfmarks. 
### We can drop these columns in the following way.
### Alternatively you can use:  df.drop(columns=to_drop, inplace=True)

In [9]:
to_drop = ['Edition Statement',
            'Corporate Author',
            'Corporate Contributors',
            'Former owner',
            'Engraver',
            'Contributors',
            'Issuance type',
            'Shelfmarks']

df.drop(to_drop, inplace=True, axis=1)

In [10]:
df.head(5)

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

### 2. Changing the Index of a DataFrame

### A Pandas Index extends the functionality of NumPy arrays to allow for more versatile slicing and labeling. 
### In many cases, it is helpful to use a uniquely valued identifying field of the data as its index.

### For example, in the dataset used in the previous section, it can be expected that when a librarian searches for a record, they may input the unique identifier (values in the Identifier column) for a book.

In [12]:
df.loc[0]

Identifier                                                            206
Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 0, dtype: object

In [13]:
df['Identifier'].is_unique

True

### Replace the existing index with this column using set_index.

In [14]:
df = df.set_index('Identifier')
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

### We can access each record in a straightforward way with loc[]. Although loc[] may not have all that intuitive of a name, it allows us to do label-based indexing, which is the labeling of a row or record without regard to its position.

In [15]:
df.loc[206]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

### 206 is the first label of the index. To access it by position, we could use df.iloc[0], which does position-based indexing.

In [16]:
df.iloc[0]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

# 3. Tidying up Fields in the Data

### Now clean specific columns and get them to a uniform format to get a better understanding of the dataset and enforce consistency. In particular, we will be cleaning 'Date of Publication' and 'Place of Publication'.

### Upon inspection, all of the data types are currently the object dtype, which is roughly analogous to str in native Python.

### It encapsulates any field that can’t be neatly fit as numerical or categorical data. 

In [9]:
df.dtypes

Place of Publication    object
Date of Publication     object
Publisher               object
Title                   object
Author                  object
Flickr URL              object
dtype: object

### Date of publication can be enforced to numeric
### 

In [26]:
df.head(50)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London         1879 [1878]   
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   
1280                                   Coventry                1802   
1808                                Christiania                1859   
1905                                    Firenze                1888   
1929                                  Amsterdam         1839, 38-54   
2836                                     Savona                1897   
2854                                     London                1865   
2956                                      Paris             1860-63   
2957                                      Paris                1873   
3017                                Puerto-Rico                1866   
3131                                   New York                1899   
4598                                       Hull                1814   
4884                                     London                1820   
4976                                     Oxonii                1800   
5382                                     London  1847, 48 [1846-48]   
5385                                     London             [1897?]   
5389                                     London             [1897?]   
5432                                     Milano                1893   
6036                                     London                1805   
6821                                   Aberdeen                1837   
7521                                       Wien                1896   
7630                                   Abingdon                1898   
8239                                     Quebec                1899   
8435                                     London                1892   
8440                                     London                1894   
11361                                   Leipzig            [1894-96   
11852        pp. 40. W. Cann: Plymouth, [1876?]                 NaN   
13074                                    London                1899   
13364                                    London              [1885]   
14466                                    London             [1860?]   
14703                                      1845                1846   
15141                                    London                1893   
15146                                    London                1894   
16173                                       enk                1817   
16543                                    London                1816   
16544                                    Boston                1817   
17602                                    London                1897   
17752                                   Upsaliæ              [1833]   
18173                                    London                1804   
18218                                 Edinburgh                1859   
18347                                   Yverdon                1777   

                                                    Publisher  \
Identifier                                                      
206                                          S. Tinsley & Co.   
216                                              Virtue & Co.   
218                                     Bradbury, Evans & Co.   
472      

In [18]:
df.loc[1905:,'Date of Publication']

Identifier
1905              1888
1929       1839, 38-54
2836              1897
2854              1865
2956           1860-63
              ...     
4158088           1838
4158128       1831, 32
4159563      [1806]-22
4159587           1834
4160339        1834-43
Name: Date of Publication, Length: 8275, dtype: object

### A particular book can have only one date of publication. Therefore, we need to do the following:

###    Remove the extra dates in square brackets, wherever present: 1879 [1878]
###    Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54
###    Completely remove the dates we are not certain about and replace them with NumPy’s NaN: [1897?]
###    Convert the string nan to NumPy’s NaN value

### Synthesizing these patterns, we can actually take advantage of a single regular expression to extract the publication year:

### The 'r' in front tells Python the expression is a raw string. In a raw string, escape sequences are not parsed. For example, '\n' is a single newline character. But, r'\n' would be two characters: a backslash and an 'n'.
### If expand=False and pat has only one capture group, then return a Series (if subject is a Series) or Index (if subject is an Index). Returns all matches (not just the first match). 

In [20]:
regex = r'^(\d{4})'
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extr.head()


Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [21]:
df['Date of Publication'].dtype
extr.dtype

dtype('O')

### Technically, this column still has object dtype, but we can easily get its numerical version with pd.to_numeric

In [22]:
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

### Some values being missing, which is a small
### price to pay for now being able to do computations on the remaining valid values:

In [23]:
df['Date of Publication'].isnull().sum() / len(df)

0.11717147339205986

### 4. Combining str Methods with NumPy to Clean Columns

### Above, you may have noticed the use of df ['Date of Publication'].str. This attribute is a way to access speedy string operations in Pandas that largely mimic operations on native Python strings or compiled regular expressions, such as .split(), .replace(), and .capitalize().

### To clean the Place of Publication field, we can combine Pandas str methods with NumPy’s np.where function.


In [35]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

### We see that for some rows, the place of publication is surrounded by other unnecessary information. If we were to look at more values, we would see that this is the case for only some rows that have their place of publication as ‘London’ or ‘Oxford’. 

In [36]:
df.loc[4157862]

Place of Publication                                  Newcastle-upon-Tyne
Date of Publication                                                  1867
Publisher                                                      T. Fordyce
Title                   Local Records; or, Historical Register of rema...
Author                      FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4157862, dtype: object

In [37]:
df.loc[4159587]

Place of Publication                                  Newcastle upon Tyne
Date of Publication                                                  1834
Publisher                                                Mackenzie & Dent
Title                   An historical, topographical and descriptive v...
Author                                              Mackenzie, E. (Eneas)
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4159587, dtype: object

### These two books were published in the same place, but one has hyphens in the name of the place while the other does not.

### To clean this column in one sweep, we can use str.contains() to get a boolean mask.

### We clean the column as follows:

In [24]:
pub = df['Place of Publication']
london = pub.str.contains('London')
london[:10]

Identifier
206      True
216      True
218      True
472      True
480      True
481      True
519      True
667     False
874      True
1143     True
Name: Place of Publication, dtype: bool

In [25]:
oxford = pub.str.contains('Oxford')
df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               pub.str.replace('-', ' ')))

In [27]:
df['Place of Publication'].head(20)

Identifier
206          London
216          London
218          London
472          London
480          London
481          London
519          London
667          Oxford
874          London
1143         London
1280       Coventry
1808    Christiania
1905        Firenze
1929      Amsterdam
2836         Savona
2854         London
2956          Paris
2957          Paris
3017    Puerto Rico
3131       New York
Name: Place of Publication, dtype: object

### 5. Renaming Columns and Skipping Rows

### Often, the datasets you’ll work with will have either column names that are not easy to understand, or unimportant information in the first few and/or last rows, such as definitions of the terms in the dataset, or footnotes.

### In that case, we’d want to rename columns and skip certain rows so that we can drill down to necessary information with correct and sensible labels.

### To demonstrate how we can go about doing this, let’s first take a glance at the initial five rows of the “olympics.csv” dataset:

In [28]:
olympics_df = pd.read_csv('/home/lenovo/Documents/Course Material 2020-21/Data Analytics 2020-21/Datasets for DA/olympics.csv')

In [29]:
olympics_df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

### This is messy indeed! The columns are the string form of integers indexed at 0. The row which should have been our header (i.e. the one to be used to set the column names) is at olympics_df.iloc[0]. This happened because our CSV file starts with 0, 1, 2, …, 15.

### Also, if we were to go to the source of this dataset, we’d see that NaN above should really be something like “Country”, ? Summer is supposed to represent “Summer Games”, 01 ! should be “Gold”, and so on.

### Therefore, we need to do two things:

    #### Skip one row and set the header as the first (0-indexed) row
    #### Rename the columns

### We can skip rows and set the header while reading the CSV file by passing some parameters to the read_csv() function.

### This function takes a lot of optional parameters, but in this case we only need one (header) to remove the 0th row:

In [41]:
olympics_df = pd.read_csv('/home/lenovo/Documents/Course Material 2020-21/Data Analytics 2020-21/Datasets for DA/olympics.csv', header=1)

In [42]:
olympics_df.head()

Unnamed: 0  ? Summer  01 !  02 !  03 !  Total  ? Winter  \
0        Afghanistan (AFG)        13     0     0     2      2         0   
1            Algeria (ALG)        12     5     2     8     15         3   
2          Argentina (ARG)        23    18    24    28     70        18   
3            Armenia (ARM)         5     1     2     9     12         6   
4  Australasia (ANZ) [ANZ]         2     3     4     5     12         0   

   01 !.1  02 !.1  03 !.1  Total.1  ? Games  01 !.2  02 !.2  03 !.2  \
0       0       0       0        0       13       0       0       2   
1       0       0       0        0       15       5       2       8   
2       0       0       0        0       41      18      24      28   
3       0       0       0        0       11       1       2       9   
4       0       0       0        0        2       3       4       5   

   Combined total  
0               2  
1              15  
2              70  
3              12  
4              12

### We now have the correct row set as the header and all unnecessary rows removed. Take note of how Pandas has changed the name of the column containing the name of the countries from NaN to Unnamed: 0.

### To rename the columns, we will make use of a DataFrame’s rename() method, which allows you to relabel an axis based on a mapping (in this case, a dict).

### Let’s start by defining a dictionary that maps current column names (as keys) to more usable ones (the dictionary’s values).
### Setting inplace to True specifies that our changes be made directly to the object. 

In [43]:
new_names =  {'Unnamed: 0': 'Country',
               '? Summer': 'Summer Olympics',
               '01 !': 'Gold',
               '02 !': 'Silver',
               '03 !': 'Bronze',
               '? Winter': 'Winter Olympics',
               '01 !.1': 'Gold.1',
               '02 !.1': 'Silver.1',
               '03 !.1': 'Bronze.1',
               '? Games': '# Games',
               '01 !.2': 'Gold.2',
               '02 !.2': 'Silver.2',
               '03 !.2': 'Bronze.2'}

In [44]:
olympics_df.rename(columns=new_names, inplace=True)

In [45]:
olympics_df.head()

Country  Summer Olympics  Gold  Silver  Bronze  Total  \
0        Afghanistan (AFG)               13     0       0       2      2   
1            Algeria (ALG)               12     5       2       8     15   
2          Argentina (ARG)               23    18      24      28     70   
3            Armenia (ARM)                5     1       2       9     12   
4  Australasia (ANZ) [ANZ]                2     3       4       5     12   

   Winter Olympics  Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
0                0       0         0         0        0       13       0   
1                3       0         0         0        0       15       5   
2               18       0         0         0        0       41      18   
3                6       0         0         0        0       11       1   
4                0       0         0         0        0        2       3   

   Silver.2  Bronze.2  Combined total  
0         0         2               2  
1         2         8              15  
2        24        28              70  
3         2         9              12  
4         4         5              12